datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -333,6 +333,12 @@ fn mathematics_numerical_coercion(
         (Null, dec_type @ Decimal128(_, _)) | (dec_type @ Decimal128(_, _), Null) => {
             Some(dec_type.clone())
         }
+        (Decimal128(_, _), Float32 | Float64) => {
+            Some(Float64)
+        }
+        (Float32 | Float64, Decimal128(_, _)) => {
+            Some(Float64)
+        }
```

Comment:
Follow Spark's type coercision rule `DecimalPrecision`.

Reply:
I think we need to add the same in `comparison_binary_numeric_coercion` and that helps towards q6

Reply:
> I think we need to add the same in `comparison_binary_numeric_coercion` and that helps towards q6

the rule between arithmetic op and logical op may be different.